### ISSUE TITLE
RFC 9266: Channel Bindings for TLS 1.3 support

### ISSUE THREAD

**Issue by Neustradamus** [openssl/openssl#18893]
Can you add the support of RFC 9266: Channel Bindings for TLS 1.3?
- https://datatracker.ietf.org/doc/html/rfc9266

Little details, to know easily:
- tls-unique for TLS =< 1.2
- tls-exporter for TLS = 1.3

Thanks in advance.

**Comment by davidben** [comment]
OpenSSL already implements `SSL_export_keying_material`. Or is the request for a convenience wrapper specific to the tls-exporter channel binding?

**Comment by Neustradamus** [comment]
@davidben: It is linked to:
- https://github.com/openssl/openssl/issues/12221

**Comment by Neustradamus** [comment]
@davidben: It is possible to add in code:
- RFC 5929
- tls-unique
- RFC 9266
- tls-exporter

It will be better to find in the code...

Example:
- https://github.com/openssl/openssl/search?q=rfc

**Comment by Neustradamus** [comment]
@davidben: For example, there were improvements in GnuTLS, glib and glib-networking about rfc9266 and tls-exporter:
- https://gitlab.com/gnutls/gnutls/-/issues/1391
- https://gitlab.gnome.org/GNOME/glib/-/issues/2710
- https://gitlab.gnome.org/GNOME/glib-networking/-/issues/191

Can you expose tls-unique (rfc5929) and tls-exporter (rfc9266)?

**Comment by Neustradamus** [comment]
@davidben, @t8m: I wish you a Happy New Year 2023!

Have you looked for this improvement request?

**Comment by mvduin** [comment]
> * tls-exporter for TLS = 1.3

tls-exporter does not require TLS 1.3, it is also defined for earlier versions of TLS provided that extended master secret (RFC7627) is used and renegotiation is disabled.  Note that the exporter context is empty but not omitted (i.e. use_context=1, contextlen=0), a distinction that's irrelevant for TLS 1.3 but matters for earlier versions.

**Comment by Neustradamus** [comment]
@mvduin: Channel Bindings for TLS = 1.3 needs tls-exporter.
Channel Bindings for TLS =< 1.2 needs tls-unique.

**Comment by Neustradamus** [comment]
@mvduin: Thanks for your answer.

There is not an announcement in code: RFC5929 / RFC9266
And I do not find:
- tls-unique
- tls-server-end-point
- tls-exporter

Example GnuTLS:
- tls-unique: https://gitlab.com/search?search=tls-unique&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- tls-exporter: https://gitlab.com/search?search=tls-exporter&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- tls-server-end-point: https://gitlab.com/search?search=tls-server-end-point&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- rfc5929: https://gitlab.com/search?search=rfc5929&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- rfc9266: https://gitlab.com/search?search=rfc9266&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master

**Comment by Neustradamus** [comment]
Dear all,

I have update the main description about tls-unique, tls-server-end-point, tls-exporter and I have added XEP-0388/XEP-0440/XEP-0474 links.

I think that you have seen the jabber.ru MITM:
- https://notes.valdikss.org.ru/jabber.ru-mitm/
- https://snikket.org/blog/on-the-jabber-ru-mitm/
- https://www.devever.net/~hl/xmpp-incident
- https://blog.jmp.chat/b/certwatch

**Comment by mvduin** [comment]
> There is not an announcement in code: RFC5929 / RFC9266 And I do not find:
> 
> * tls-unique
> * tls-server-end-point
> * tls-exporter

Sorry, I really have no idea what you're trying to say or what I'm supposed to notice in those linked searches. [RFC9266](https://www.rfc-editor.org/rfc/rfc9266.html) explicitly says that:
* `tls-unique` is NOT supported by TLS 1.3
* `tls-unique` remains the default for TLS 1.2 for purposes such as SCRAM or GSS-API over SASL
* `tls-exporter` replaces `tls-unique` as default binding mechanism for TLS 1.3 in aforementioned applications
* `tls-exporter` MUST be supported by any implementation that supports channel binding with TLS 1.3
* `tls-exporter` is also supported with older versions of TLS as long as certain conditions are met (renegotiation is disabled, extended master secret has been negotiated)

RFC 9266 is silent about `tls-server-end-point` so it appears it's not defined for TLS 1.3, unless there's some other RFC that I've overlooked. Now, I don't see an obvious reason why you wouldn't be able to use it with TLS 1.3 if you accept the intrinsic limitations of this channel binding type, but nevertheless any such usage would be non-standard and at your own risk.

Having said that, if I were in a situation where my only choices are using `tls-server-end-point` binding with TLS 1.3 (despite not being officially supported) versus not using channel binding at all and just hoping users will pay attention to never use their credentials to authenticate to untrustworthy domains, I would unhesitatingly choose the former.

**Comment by SamWhited** [comment]
> RFC 9266 is silent about tls-server-end-point so it appears it's not defined for TLS 1.3, unless there's some other RFC that I've overlooked

My apologies for being vague there; you are correct that `tls-exporter` is not defined for TLS 1.3 (per [RFC 8446 appendix C.5](https://www.rfc-editor.org/rfc/rfc8446#appendix-C.5)). It should no longer be considered secure for use more generally as well as it suffers from many of the same uniqueness problems as tls-unique without the master secret fix, so you have to be absolutely sure that everything is configured correctly.

**Comment by Neustradamus** [comment]
@mvduin: Thanks for your answer.

Yes, tls-unique is for TLS =< 1.2, It is already noted in main description.

RFC9266 says that tls-unique is replaced by tls-exporter with TLS 1.3 but do not change about tls-server-end-point.
`
Furthermore, this document updates [RFC5929] by adding an additional unique channel binding type, "tls-exporter", that replaces some usage of "tls-unique".
`

For example, when you search in GnuTLS code, you can find all easily, there are references:
- tls-unique: https://gitlab.com/search?search=tls-unique&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- tls-exporter: https://gitlab.com/search?search=tls-exporter&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- tls-server-end-point: https://gitlab.com/search?search=tls-server-end-point&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- rfc5929: https://gitlab.com/search?search=rfc5929&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master
- rfc9266: https://gitlab.com/search?search=rfc9266&nav_source=navbar&project_id=179611&group_id=121613&search_code=true&repository_ref=master

Same in openssl code, nothing, there is no reference, can you add in code to help people?
- tls-unique: https://github.com/search?q=repo%3Aopenssl%2Fopenssl+tls-unique&type=code
- tls-exporter: https://github.com/search?q=repo%3Aopenssl%2Fopenssl+tls-exporter&type=code
- tls-server-end-point: https://github.com/search?q=repo%3Aopenssl%2Fopenssl+tls-server-end-point&type=code
- rfc5929: https://github.com/search?q=repo%3Aopenssl%2Fopenssl+rfc5929&type=code
- rfc9266: https://github.com/search?q=repo%3Aopenssl%2Fopenssl+rfc9266&type=code

**Comment by mvduin** [comment]
> > RFC 9266 is silent about tls-server-end-point so it appears it's not defined for TLS 1.3, unless there's some other RFC that I've overlooked
> 
> My apologies for being vague there; you are correct that `tls-exporter` is not defined for TLS 1.3 (per [RFC 8446 appendix C.5](https://www.rfc-editor.org/rfc/rfc8446#appendix-C.5)). It should no longer be considered secure for use more generally as well as it suffers from many of the same uniqueness problems as tls-unique without the master secret fix, so you have to be absolutely sure that everything is configured correctly.

???  I'm assuming you meant to say `tls-server-end-point` rather than `tls-exporter` (which was introduced specifically as a mechanism that supports TLS 1.3), but even then I don't think this is true. `tls-server-end-point` binding doesn't provide the same security properties as `tls-unique`/`tls-exporter` but its limitations are not similar to the issues of `tls-unique`. And importantly, `tls-server-end-point` binding can be used in situations where other binding mechanisms are infeasible, e.g. where TLS is being terminated by a reverse proxy, since it doesn't require the authentication endpoint to have access to any TLS secrets, it only needs to know what server certificate was used.

In terms of security I'd still greatly prefer `tls-server-end-point` channel binding over just sending your username and password to the server and hope the user doesn't get phished, which is _by far_ the most common authentication protocol used today :P

**Comment by SamWhited** [comment]
Oops, yes indeed, `tls-server-end-point`.

**Comment by mvduin** [comment]
> It should no longer be considered secure for use more generally as well as it suffers from many of the same uniqueness problems as tls-unique without the master secret fix

I've had a moment to look up the triple handshake paper to refresh my memory on how the attack worked, and in the scenario where lack of extended master secret allows `tls-exporter` and `tls-unique` to be broken (due to both ends of a MITM attack ending up with the same master secret and, after resume, Finished handshake, which means the attacker can simply forward authentication messages), `tls-server-end-point` is unaffected since it's based purely on the server certificate which means the client's authenticator would incoporate the attacker's certificate and therefore be rejected by the legitimate server.

**Comment by SamWhited** [comment]
I went back and refreshed myself on this as well, you're correct, `tls-server-end-point` would have prevented the recent attack that the previous poster was mentioning for TLS 1.2. It's weaknesses lie more in the fact that it's using a static secret so if the cert gets stolen the MITM can't be detected (where `unique` type channel bindings such as `tls-unique` and `tls-exporter` would still detect the MITM since they use parts of the TLS session).

Anyways, sorry for the confusion, point is that `tls-server-end-point` is not defined for TLS 1.3, though now I can't remember why off the top of my head. :)

**Comment by mvduin** [comment]
Indeed. I'd love to know the reason it wasn't defined for TLS 1.3, if there actually is one that's not "you ought to be using `tls-exporter` instead" (which ignores the situations where removing the option to use `tls-server-end-point` isn't going to encourage use of `tls-exporter` but instead will lead to abandoning channel binding entirely)

**Comment by SamWhited** [comment]
> , if there actually is one that's not "you ought to be using tls-exporter instead"

I only wrote RFC 9266 recently but TLS 1.3 had been out for a while, so this was not the reason. Sadly I recently purged my old mailing list archives to save space, but I know there was a lot of discussion on the TLS WG list at the time about this, so I'm sure you could go dig through the archives and find it.

**Comment by mvduin** [comment]
> I only wrote RFC 9266 recently

OHH! I hadn't noticed your name is also the one at the top of that RFC! Greetings!  Suddenly the "My apologies for being vague there" makes sense, I didn't quite get that at first. But indeed, it seems a bit odd the RFC is just completely silent about it instead of either defining it for TLS 1.3 or explaining why it isn't.

Renegotiation is of course a hazard with TLS 1.2 but that's of course irrelevant as argument for not defining `tls-server-end-point` for TLS 1.3 which doesn't have renegotiation.

**Comment by SamWhited** [comment]
> it seems a bit odd the RFC is just completely silent about it instead of either defining it for TLS 1.3 or explaining why it isn't.

Yes, sorry, in retrospect this was an oversight on my part and I should have at least mentioned why I didn't define one in the document. The simple answer is "that's not the document I was writing, it's just out of scope, someone else can do that if they want" :)

> Renegotiation is of course a hazard with TLS 1.2 but that's irrelevant as argument for not defining tls-server-end-point for TLS 1.3 which doesn't have renegotiation.

Not necessarily; TLS 1.3 would have required a lot of work to *prove* that it did not suffer from the same or similar flaws. There may not be renegotiation, but there are a lot of other extensions and changes which could introduce similar issues (eg. 0-RTT TLS, new hashes and ciphers, etc.) and it would have required a lot more work and careful consideration to check if the older channel binding methods were secure and under what conditions they were secure. This is (I suspect) why the TLS WG just said "these aren't defined" (perhaps with an implicit "for now").

As previously mentioned I only defined a unique one and not an endpoint one because that was just more work and out of scope for the problem I wanted to solve. In general I think that endpoint based methods have less value than unique (session) based methods for reasons previously stated. That's not to say they have *no* value, I just didn't think it was worth my time to define both. Though bear in mind that I am certainly no expert here, I wrote this document with a lot of help from the WG and don't want to position myself as having more knowledge than I do. It's quite possible I'm missing some obvious and major benefit to endpoint based channel binding.
All that being said, I'm sure the KITTEN WG wouldn't turn *down* a valid endpoint method for TLS 1.3 if someone wanted to write one or prove that `tls-server-end-point` was secure, it's just not something I was interested in working on as the security properties aren't as robust in general.

But we're veering fairly far from the point of this issue; maybe we could take this OOB unless this conversation has value to others on this thread? My email and JID are the same and can be found in the RFC; feel free to reach out any time.

**Comment by mvduin** [comment]
Fair enough, and you're absolutely right it got way off topic, my apologies!

**Comment by hlandau** [comment]
To reiterate, my understanding of the situation is that we already have the relevant exporter API in OpenSSL. It is enough to implement the channel binding specification, which is a fairly trivial construction over exporters.

I therefore don't think this is a massive gap in capability that requires prioritisation in terms of code added to OpenSSL. However, a convenience function which implements this RFC would probably still be welcome as a PR and if approved eligible for the next feature release.

**Comment by Neustradamus** [comment]
@t8m: Thanks for your recent changes about:
- https://github.com/openssl/openssl/issues/12221

Do you work on this ticket? :)

It will be important to have a perfect support and commented source code (to easy understand) about:
- tls-unique for TLS =< 1.2 (RFC5929) / Note: TLS 1.2 is very used
- tls-server-end-point =< 1.2 + 1.3 (RFC5929)
- tls-exporter for TLS = 1.3 (RFC9266)

Commented source code about terms here:
- https://github.com/openssl/openssl/issues/18893#issuecomment-1203315748

For example: GnuTLS commented source code:
- https://github.com/openssl/openssl/issues/18893#issuecomment-1793602450

For example, some global GitHub searches:
- https://github.com/search?q=rfc9266&type=code
- https://github.com/search?q=%22rfc+9266%22&type=code
- https://github.com/search?q=rfc5929&type=code
- https://github.com/search?q=%22rfc+5929%22&type=code
- https://github.com/search?q=tls-exporter&type=code
- https://github.com/search?q=tls-server-end-point&type=code
- https://github.com/search?q=tls-unique&type=code
- ...

Thanks a lot in advance.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
